colander/__init__.py

```diff
@@ -1838,20 +1838,19 @@ def deserialize(self, node, cstruct):
             result = iso8601.parse_date(cstruct)
             result = result.time()
         except (iso8601.ParseError, TypeError):
+            fmt = {
+                8:'%H:%M:%S',
+                5:'%H:%M'
+            }.get(len(cstruct), '%H:%M:%S.%f')
```

Comment:
Instead of trying to guess based upon length, why not just attempts to go through each one in order? Start with the ms one, followed by the one with seconds, followed by the one with just hours:minutes?

Much like the original solution did, but loop over a list of formats.

Reply:
I would also prefer a solution that looped over the formats and tried them. Finally if none of them worked then raise an exception. You could pass in the final error object as the `err` key. I don't think any nested exceptions would occur if this were a loop.

I'm concerned about testing the length because I don't actually know if the `%H` etc always maps to two characters - it defers to libc and possibly any locale settings which may cause unexpected regressions. It's safer to just test the formats and use the one that works since the formats are unambiguous.